feat: add private-to-private docs (#319)
libp2p/docs: content/concepts/transports/webrtc.md

{{< /alert >}}

For this transport, libp2p supports its own signaling protocol which has the protocol id: `webrtc-signaling`.

#### How private-to-private connectivity works

<!-- TO ADD DIAGRAM -->

Suppose we have three network entities:

- _Node A_ - a libp2p node running in the browser
- _Node B_ - another browser libp2p node running on a different browser instance
- _Relay R_ - a libp2p relay node

In this connectivity scenario, _A_ wants to connect to _B_.
This works as follows:

- _B_ connects to _R_ and makes a relay reservation. It appends `/webrtc` to its relayed multiaddress and advertises it to the network.
- _A_ discovers _B_'s relayed multiaddr, sees it supports private-to-private, and establishes a relayed connection to _B_ via _R_.
- _A_ and _B_ create outbound and inbound [`RTCPeerConnection` objects](https://webrtc.org/getting-started/peer-connections) respectively
  - Per their namesake, these objects handle peer connections, define how they're set up, and contain other useful information about the network.
- _A_ initiates the `webrtc-signaling` protocol to _B_ via a stream over the relayed connection
  - This signaling stream is used to:

>     - Exchange SDPs between nodes (an offer from _A_ and an answer from _B_)
>     - Exchange information about the network connection (in WebRTC parlance this is called [exchanging ICE candidates](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Connectivity#ice_candidates))

Author: ```suggestion
    - Exchange SDPs between nodes (an offer from _A_ and an answer from _B_)
      - This include exchangning information about the network connection (in WebRTC parlance this is called [exchanging ICE candidates](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Connectivity#ice_candidates))
```

>     - Exchange information about the network connection (in WebRTC parlance this is called [exchanging ICE candidates](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Connectivity#ice_candidates))

Reviewer: The ICE candidates are part of the SDP, thus is a duplicate of the bullet above.

Author: I would like to keep this, made it a subpoint of the previous line.

> - After this whole process is done, there can be one of two results:

Reviewer: ```suggestion
- The SDP messages are passed to the browsers WebRTC stack, which then tries to establish a direct connection.
```

  - A successful direct connection is established between _A_ and _B_
    - In this case, both browser nodes will close the signaling protocol stream
    - The relayed connection is closed
  - A failed direction connection
    - In this case, the signaling stream is reset

>     - It's important to note that libp2p does not specify transferring data over the relayed connection if the direct connection fails

Author: ```suggestion
```

>     - It's important to note that libp2p does not specify transferring data over the relayed connection if the direct connection fails

Reviewer: Is this really necessary for newcomers to know?

Author: Removed.

<!-- TO ADD DIAGRAM -->

### WebRTC W3C (Hole punching to achieve Browser-to-Browser connectivity)

> Thanks to js-libp2p and rust-libp2p (complied to Wasm), libp2p can run in the browser environment.

Reviewer 2: We don't use the wasm compilation for browser to browser, nor elsewhere, do we? Consider removing?
```suggestion
Thanks to js-libp2p and rust-libp2p (compiled to Wasm), libp2p can run in the browser environment.
```

Author: My intention with this line was to indicate it's possible to run a rust libp2p node in the browser (when compiled to Wasm), not to indicate browser to browser is supported in rust-libp2p. I can either make this more clear or delete this.

Reviewer: It is not implemented yet. It is on the roadmap only.

https://github.com/libp2p/rust-libp2p/blob/master/ROADMAP.md#webrtc-support-browser-to-browser

However, browsers impose certain restrictions on application code (such as libp2p browser nodes).
Applications are sandboxed and face constraints on security and networking.
For instance, browsers do not permit direct access to raw network sockets.
Additionally, it's a sure bet that libp2p browser nodes will be behind a NAT/firewall.

> Due to these restrictions, browser nodes cannot listen for incoming connections,

Reviewer 2: They also can't dial TCP and QUIC connections.

and as a result, they cannot communicate with other browser nodes.

> Thankfully, libp2p solves this problem and enables browser node to browser node connectivity by supporting a transport called WebRTC W3C.

Reviewer 2: Why W3C? All the W3C is does is specify the browser API. We don't call WebTransport "WebTransport W3C".
```suggestion
Thankfully, libp2p solves this problem and enables browser node to browser node connectivity by supporting WebRTC.
```

Author: Leaving this until the convo in specs is resolved.

{{< alert icon="" context="info">}}

> Due to the restrictions mentioned above, libp2p browser nodes are private nodes (meaning they cannot be dialed from outside of their local network).

Reviewer 2: Not really. They're private nodes because people run their browsers in their home / corporate networks, and those networks almost always use NATs. In principle, you could run your browser on a machine that has a public address.

Author: I'll make this clearer

However, in principle, this use case can be thought of not just as browser-to-browser but private node to private node.

> This WebRTC solution enables connectivity between two non browser nodes behind a NAT/firewall as well

Reviewer 2: Not 100% sure if this is correct.
```suggestion
This WebRTC solution enables connectivity between two non-browser nodes behind a NAT / firewall as well
```

{{< /alert >}}

#### Transport Internals

> The libp2p WebRTC W3C transport is enabled by supporting the [W3C defined](https://w3c.github.io/webrtc-pc/#introduction) `RTCPeerConnection` API.

Reviewer 2: I don't think mentioning W3C makes a lot of sense here. I will engage on the specs PR. Browser-to-server also uses the W3C-defined API, simply because that's the API that's available in browsers.

This core API enables p2p connectivity and provides methods for establishing connections and transferring streams of data between peers.

> Running instances of libp2p that support this transport will have `/webrtc-w3c` in their multiaddr.

Reviewer 2: Same here, I don't think we should go with this multiaddr, but I'll comment on the PR in the multiaddr repo.

Reviewer 2: Commented: https://github.com/multiformats/multiaddr/pull/150#event-8661620493.

> However, there's more to p2p connections that what `RTCPeerConnection` provides. Crucially, signaling isn't built into the WebRTC API.

Reviewer 2: It kind of is. The W3C gives you the signaling messages, you just need to figure out how to ship them to the peer yourself.

Author: I see, might have misrepresented what was described [here](https://web.dev/webrtc-basics/#signaling-session-control,-network,-and-media-information): "Signaling methods and protocols are not specified by WebRTC. Signaling is not part of the RTCPeerConnection API."

I'll take a note out of this: https://webrtc.org/getting-started/peer-connections#signaling and reword as:
"Crucially, the WebRTC specification doesn't prescribe any single solution for signaling. The peer connection API has a notion of [signaling state](https://w3c.github.io/webrtc-pc/#dom-rtcsignalingstate) but not mechanism to share this data with remote peers."

lmk what you think

> However, there's more to p2p connections that what `RTCPeerConnection` provides. Crucially, signaling isn't built into the WebRTC API.

Reviewer: ```suggestion
However, there's more to p2p connections than what `RTCPeerConnection` provides. Crucially, signaling isn't built into the WebRTC API.
```

{{< alert icon="" context="info">}}
Signaling is the process of coordinating communication and exchanging metadata about the communication (i.e. initializing, closing, or reporting errors about connections).
{{< /alert >}}

For this transport, libp2p supports its own signaling protocol which has the protocol id: `webrtc-w3c-signaling`.

#### How browser-to-browser connectivity works

<!-- TO ADD DIAGRAM -->

Suppose we have three network entities:

- _Node A_ - a libp2p node running in the browser
- _Node B_ - another browser libp2p node running on a different browser instance
- _Relay R_ - a libp2p relay node

In this connectivity scenario, _A_ wants to connect to _B_.

> This roughly works as follows:

Reviewer: Is this step by step documentation needed? To me this is just rephrasing what is written in the spec without being higher level, thus more helpful to newcomers.

Author: Good point, I initially wrote the step by step to help me understand the flow but I agree it is a rehash. I rewrote it to keep it at a higher level and elide some of the details. Let me know what you think.

> - _A_ and _B_ are both connected to _R_
> - _B_ appends `webrtc-w3c` to its multiaddress and that multiaddress is relayed by _R_.
> - _A_ will discover _B_'s multiaddress via _R_
> - _A_ sees that _B_ supports `webrtc-w3c` and establishes a relayed connection to _B_

Reviewer: _B_ will most likely not advertise its address through _R_ to _A_. This is a chicken-and-egg problem. How would _A_ know to connect to _R_ without _B_'s multiaddr?

```suggestion
- _B_ connects to _R_ and makes a relay reservation. (See circuit relay v2 docs.)
- _B_ appends `webrtc-w3c` to its relayed multiaddress and advertises it to the network.
- _A_ will discover _B_'s relayed multiaddr.
- _A_ sees that _B_ supports `webrtc-w3c` and establishes a relayed connection to _B_ via _R_.
```

- Over the relayed connection, _A_ will:

>   - Create an _outbound_ `RTCPeerConnection`

Reviewer: Just to avoid confusion, one does not create an `RTCPeerConnection` over the relayed connection, but rather with the help of the relayed connection. Is the phrasing only confusing me?

>   - Send a SDP offer via the WebRTC API

Reviewer: I am confused. One can create an SDP offer via the `RTCPeerConnection` and then send that offer via the libp2p signaling stream. I don't know how one would send an SDP offer via the `RTCPeerConnection`.
